### DISCUSSION TITLE
Missing json_encoders in v2 - Any similar functionality

### DISCUSSION THREAD

**Discussion by sachingiridhar-acelr** [pydantic/pydantic#6697] [Pydantic V2]
This is just a small example of my use case. I have a RootModel which I use across my project. I have to json encode this at one place to avoid duplication of code (DRY). 
```python
from datetime import time as time_only
from pydantic import BaseModel, Field

def convert_time_to_hh_mm_ss(dt: time_only) -> str:
    return dt.strftime('%H:%M:%S')

class RootModel(BaseModel):
    class Config:
        json_encoders = {
            time_only: convert_time_to_hh_mm_ss
        }
```

I can see that json_encoders are removed in Pydantic v2. So how do I achieve the same in v2. This is very important to me since I have around 20 to 30 different models which needs this functionality

**Comment by xmatthias** [comment]
Facing a similar problem in #6726
while my problem is not directly applicable for a root-base model - it's still a quite similar problem.

**Comment by eyemyth** [comment]
I believe the new approach would be to use the [functional serializers](https://docs.pydantic.dev/latest/usage/serialization/#custom-serializers)

**Comment by samuelcolvin** [comment]
Yes, functional serializers are the proper solution, but we've just added support (albeit with a deprecation warning) back in #6811

**Comment by voxsim** [comment]
@samuelcolvin probably makes sense to have somewhere an example that translates from the v1 to v2 because it is not super clear from the documentation.

**Comment by Samreay** [comment]
I'd be keen to get an official example as well.

Our use case is fairly simple, in that our pydantic models have some pandas dataframes as attritubtes, so we have `json_encoders={pd.DataFrame=lambda x: x.reset_index().to_dict(orient="list")}`.

Given this applies to all dataframe attritbutes without having to write out the field name for all of them, its very handy.

But with the new approach, I'm not sure if I should be using `field_serializer`, `model_serialiser`, `PlainSerializer` or `WrapSerializer`.

**Reply by alex-janss** [reply]
Here's an example using PlainSerializer that i used.
instead of:
```python
class Foo(BaseModel):
    model_config = ConfigDict(json_encoders={str: lambda x: ' '.join(x)})
    x: str
    y: str = Field(...)
```
create a custom type:
```python
CustomStr = Annotated[str, PlainSerializer(lambda x: ' '.join(x))]
class Foo(BaseModel):
    x: CustomStr
    y: CustomStr = Field(...)
```

**Comment by cherrywins** [comment]
How eventually have you rewritten your code?

**Comment by devamin** [comment]
I got this from the test written in the feature support PR : 

```python
    class Model(BaseModel):
        x: timedelta
        y: Decimal
        z: date

        model_config = ConfigDict(
            json_encoders={timedelta: lambda v: f'{v.total_seconds():0.3f}s', Decimal: lambda v: 'a decimal'}
        )
```

**Comment by HenningDinero** [comment]
A general question: How would you use these function-serializers it to serialize a specific object e.g all datetime objects in case you have 15 fields and don't want to decorate all of them using @field_serializer (and in case I add a nother field but forget to add that to the list of fields in `@field_serializer`? We could do that before using 

`json_encoders = {datetime: lambda v: v.strftime("%d-%m-%Y")}`

Also, to allow a base-model that defines serialization of datetime-objects that I can inherit from and use "out of the box"

**Reply by Samreay** [reply]
I'd annotate them. We have this issue already using `avro`, which represents datetimes as int format timestamps. Here's our generic way of doing it which caters to different formats sending through either datetimes, strs (either iso or str'd ints, and ints), to get a consistent UTC datetime object at the end that saves out to milliseconds epoch.


```python

def _validate_int_datetime(x: dt | int) -> dt:
    if isinstance(x, str):
        try:
            x = int(x)
        except ValueError:
            x = dt.fromisoformat(x)
    if isinstance(x, int):
        x = dt.fromtimestamp(x / 1000, tz=tz.utc)
    if x.tzinfo is None:
        x = x.replace(tzinfo=tz.utc)
    return x.astimezone(tz.utc)

IntDatetime = Annotated[
    dt,
    BeforeValidator(_validate_int_datetime),
    PlainSerializer(lambda x: int(x.timestamp() * 1000), when_used="always"),
]

class SomeModel(BaseModel):
   some_time: IntDatetime
```

**Reply by HenningDinero** [reply]
I still want the datetime-object such that I can operate on that as an usual datetime-object.
Ind the end i'll have to send some data to an API, and I have to make sure that the format is correct thus I would like to specify that when the `model.model_dump()` function is called.

Since I have plenty of different models that all follow the same structure e.g

```python
from pydantic import BaseModel

class Model(BaseModel):
   # define some general properties here
  id: str


class ModelTransaction(Model):
    transaction_time: datetime
    value: float


class ModelBooking(Model):
    booking_time: datetime
```

having to define `@field_serializer` in all of those sub-models makes it rather error-prone

**Reply by Samreay** [reply]
My example was taken from my codebase for Int serialisation. Just replace the lambda function with whatever output format you need and its good to go.

**Reply by HenningDinero** [reply]
I'll still have to annotate my datetime objects as `CustomDateTime` around in all my projects and in the future, I assume?

**Reply by Samreay** [reply]
Yeah, you'd have one annotation, preferably in a common repo / package that all your projects share, and then you would just

```python
from shared_utils.types import CustomDateTime

class SomeModel(BaseModel):
    some_time: CustomDateTime
```

Unless you wanted to specify the annotation in each repository. We have a shared repo filled with data definitions that allow different services in different languages to communicate with shared object structure,

**Reply by HenningDinero** [reply]
Indeed - but again, if I for some reason forget that and just use `datetime` then errors occurs.
That is what I'm missing; that I don't have to define other variables etc. but just use the good old `json_encoders` with a `datetime` object - that would handle it all in one go :) 

I'll have a look at your solution meanwhile though!

**Comment by ahmedelq** [comment]
I find this easy and simply: 
```python
class Person(BaseModel):
    id: uuid.UUID = Field(default_factory=uuid.uuid4)

    @field_serializer("id", when_used="always")
    def serialize_dt_offset(self, id: UUID) -> str:
        return str(id)
```
